server/src/test/java/io/spine/server/EnvSettingTest.java

```diff
@@ -222,6 +246,135 @@ private void testLazyForEnv(Class<? extends EnvironmentType<?>> type) {
         }
     }
 
+    @Nested
+    @DisplayName("be thread-safe")
+    class ThreadSafety {
+
+        private ExecutorService readWriteExecutors;
+        private CountDownLatch latch;
+        private EnvSetting<UUID> setting;
+
+        @BeforeEach
+        void setUp() {
+            readWriteExecutors = Executors.newFixedThreadPool(3);
+            latch = new CountDownLatch(1);
+            setting = new EnvSetting<>();
+        }
+
+        @Test
+        @DisplayName("allowing multiple threads to read simultaneously " +
+                "without affecting the stored value")
+        void testReadOperations() {
+            UUID initialValue = randomUUID();
+            setting.use(initialValue, Local.class);
+
+            Future<?> readBlockingFuture = runBlockingReadOperation(Local.class, initialValue);
+            sleepUninterruptibly(100, MILLISECONDS);
+
+            UUID actualValue = setting.value(Local.class);
+            assertThat(actualValue).isEqualTo(initialValue);
+
+            latch.countDown();
+            awaitFutureCompletion(readBlockingFuture);
+        }
+
+        @Test
+        @DisplayName("allowing a write operation to holds exclusive access, " +
+                "blocking concurrent reads and writes until complete")
+        void testWriteOperations() {
+            UUID initialValue = randomUUID();
+            Future<?> writeBlockingFuture = runBlockingWriteOperation(Local.class, initialValue);
+            sleepUninterruptibly(100, MILLISECONDS);
+
+            UUID rewrittenValue = randomUUID();
+            Future<?> writeFuture =
+                    runVerifyingWriteOperation(Local.class, rewrittenValue, initialValue);
+            sleepUninterruptibly(100, MILLISECONDS);
+
+            latch.countDown();
+            awaitFutureCompletion(writeFuture);
+            awaitFutureCompletion(writeBlockingFuture);
+
+            assertThat(setting.value(Local.class)).isEqualTo(rewrittenValue);
+        }
+
+        @AfterEach
+        void tearDown() {
+            try {
+                readWriteExecutors.shutdownNow();
+                readWriteExecutors.awaitTermination(500, MILLISECONDS);
+            } catch (InterruptedException e) {
+                throw illegalStateWithCauseOf(e);
```

Comment:
I am not sure we want that, I would just ignore. 

If you still think we need to re-throw, please have a method-level comment explaining the behaviour.